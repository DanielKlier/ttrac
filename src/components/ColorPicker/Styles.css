.popover.color-picker-overlay {
    max-width: 346px;
}

.color-picker .color-button, .color-picker-overlay .color-button {
    width: 24px;
    height: 24px;
}

.color-picker .color-button-outer,
.color-picker-overlay .color-button-outer {
    margin: 0 6px 6px 0;
    float: left;
}

.color-picker .color-button-outer:nth-of-type(8n),
.color-picker-overlay .color-button-outer:nth-of-type(8n) {
    margin-right: 0;
}

.color-picker .color-button-outer:last-of-type,
.color-picker-overlay .color-button-outer:last-of-type {
    margin: 0;
}

.color-picker .color-button-outer.selected .btn,
.color-picker-overlay .color-button-outer.selected .btn {
    border: 1px solid #337ab7;
    box-shadow: 0 0 1px #337ab7;
}

.color-picker .btn,
.color-picker-overlay .btn {
    padding: 4px;
}
