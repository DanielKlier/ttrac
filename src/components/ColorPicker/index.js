import * as React from 'react';
import {findDOMNode} from 'react-dom';
import PropTypes from 'prop-types';
import {Button, Overlay, Popover} from 'react-bootstrap';
import uuid from 'uuid';
import {noop} from 'lodash';
import './Styles.css';

class ColorPicker extends React.Component {

    static propTypes = {
        value: PropTypes.string,
        colorPalette: PropTypes.arrayOf(PropTypes.string),
        onChange: PropTypes.func
    };

    static defaultProps = {
        onChange: noop,
        value: '',
        colorPalette: []
    };

    constructor() {
        super();

        this.id = 'ColorPicker_' + uuid();

        this.toggleButton = null;

        this.state = {
            isOpen: false
        };
    }

    render() {
        const backgroundStyle = {
            backgroundColor: this.props.value
        };

        const colors = this.renderColors();

        return (
            <div className="color-picker">
                <Button ref={ref => this.toggleButton = ref} onClick={this.toggleOverlay}>
                    <div className="color-button" style={backgroundStyle}/>
                </Button>

                <Overlay show={this.state.isOpen} onHide={() => this.toggleOverlay(false)} rootClose
                         target={props => findDOMNode(this.toggleButton)} placement="bottom">
                    {colors}
                </Overlay>
            </div>
        );
    }

    renderColors() {
        const {value} = this.props.value;

        return (
            <Popover id={this.id} className="color-picker">
                {this.props.colorPalette.map((color, index) => {
                    const classes = [
                        'color-button'
                    ];

                    if (color === value) {
                        classes.push('selected');
                    }

                    return (
                        <Button key={index} onClick={() => this.updateColor(color)}>
                            <div className={classes.join(' ')} style={{backgroundColor: color}}/>
                        </Button>
                    );
                })}
            </Popover>
        );
    }

    updateColor = (color) => {
        this.props.onChange(color);
        this.toggleOverlay(false);
    };

    toggleOverlay = (isOpen) => {
        // Toggle semantics
        if (isOpen === undefined) {
            isOpen = !this.state.isOpen;
        }

        this.setState({isOpen});
    };
}

export default ColorPicker;
