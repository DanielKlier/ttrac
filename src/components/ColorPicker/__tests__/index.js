import React from 'react';
import {mount} from 'enzyme';
import toJSON from 'enzyme-to-json';
import ColorPicker from '../index';

const colors = [
    '#ff0000',
    '#00ff00',
    '#0000ff',
    '#ffff00',
    '#00ffff',
    '#ff00ff'
];

test('it renders just the color button when not expanded', () => {
    const component = mount(<ColorPicker colorPalette={colors} value={'#ff0000'}/>);

    expect(toJSON(component)).toMatchSnapshot();
});

test('it opens and renders the overlay when clicked', () => {
    const component = mount(<ColorPicker colorPalette={colors} value={'#ff0000'}/>);

    component.find('Button').simulate('click');

    expect(toJSON(component)).toMatchSnapshot();
});

test('it calls the change handler when a color button in the popover is clicked', () => {
    const changeHandler = jest.fn();

    const component = mount(<ColorPicker colorPalette={colors} value={'#ff00ff'}
                                         onChange={changeHandler}/>);

    component.find('Button').simulate('click');

    component.find('.color-picker-overlay').find('Button').first().simulate('click');

    expect(changeHandler).toBeCalledWith('#ff0000');

    component.update();
});
