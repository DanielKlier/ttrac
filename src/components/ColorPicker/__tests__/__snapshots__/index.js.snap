// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it opens and renders the overlay when clicked 1`] = `
<ColorPicker
  colorPalette={
    Array [
      "#ff0000",
      "#00ff00",
      "#0000ff",
      "#ffff00",
      "#00ffff",
      "#ff00ff",
    ]
  }
  onChange={[Function]}
  value="#ff0000"
>
  <div
    className="color-picker"
  >
    <div
      className="clearfix toggle-wrap"
    >
      <ColorButton
        color="#ff0000"
        disabled={false}
        isSelected={false}
        onClick={[Function]}
      >
        <div
          className="color-button"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <div
                className="color-tile"
                style={
                  Object {
                    "backgroundColor": "#ff0000",
                  }
                }
              />
            </button>
          </Button>
        </div>
      </ColorButton>
    </div>
    <Overlay
      animation={[Function]}
      onHide={[Function]}
      placement="bottom"
      rootClose={true}
      show={true}
      target={[Function]}
    >
      <Overlay
        onHide={[Function]}
        placement="bottom"
        rootClose={true}
        show={true}
        target={[Function]}
        transition={[Function]}
      >
        <Portal>
          <RootCloseWrapper
            event="click"
            onRootClose={[Function]}
          >
            <Fade
              appear={true}
              in={true}
              mountOnEnter={false}
              onExited={[Function]}
              timeout={300}
              unmountOnExit={false}
            >
              <Transition
                appear={true}
                enter={true}
                exit={true}
                in={true}
                mountOnEnter={false}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                timeout={300}
                unmountOnExit={false}
              >
                <Position
                  className="fade in"
                  containerPadding={0}
                  placement="bottom"
                  shouldUpdatePosition={false}
                  target={[Function]}
                >
                  <Popover
                    arrowOffsetLeft="NaN%"
                    bsClass="popover"
                    className="fade in color-picker-overlay"
                    id="ColorPicker_2"
                    placement="bottom"
                    positionLeft={0}
                    positionTop={0}
                    style={
                      Object {
                        "left": 0,
                        "top": 0,
                      }
                    }
                  >
                    <div
                      className="fade in color-picker-overlay popover bottom"
                      id="ColorPicker_2"
                      role="tooltip"
                      style={
                        Object {
                          "display": "block",
                          "left": 0,
                          "top": 0,
                        }
                      }
                    >
                      <div
                        className="arrow"
                        style={
                          Object {
                            "left": "NaN%",
                            "top": undefined,
                          }
                        }
                      />
                      <div
                        className="popover-content"
                      >
                        <ColorButton
                          color="#ff0000"
                          disabled={false}
                          isSelected={true}
                          key="0"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button selected"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#ff0000",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                        <ColorButton
                          color="#00ff00"
                          disabled={false}
                          isSelected={false}
                          key="1"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#00ff00",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                        <ColorButton
                          color="#0000ff"
                          disabled={false}
                          isSelected={false}
                          key="2"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#0000ff",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                        <ColorButton
                          color="#ffff00"
                          disabled={false}
                          isSelected={false}
                          key="3"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#ffff00",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                        <ColorButton
                          color="#00ffff"
                          disabled={false}
                          isSelected={false}
                          key="4"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#00ffff",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                        <ColorButton
                          color="#ff00ff"
                          disabled={false}
                          isSelected={false}
                          key="5"
                          onClick={[Function]}
                        >
                          <div
                            className="color-button"
                          >
                            <Button
                              active={false}
                              block={false}
                              bsClass="btn"
                              bsStyle="default"
                              disabled={false}
                              onClick={[Function]}
                            >
                              <button
                                className="btn btn-default"
                                disabled={false}
                                onClick={[Function]}
                                type="button"
                              >
                                <div
                                  className="color-tile"
                                  style={
                                    Object {
                                      "backgroundColor": "#ff00ff",
                                    }
                                  }
                                />
                              </button>
                            </Button>
                          </div>
                        </ColorButton>
                      </div>
                    </div>
                  </Popover>
                </Position>
              </Transition>
            </Fade>
          </RootCloseWrapper>
        </Portal>
      </Overlay>
    </Overlay>
  </div>
</ColorPicker>
`;

exports[`it renders just the color button when not expanded 1`] = `
<ColorPicker
  colorPalette={
    Array [
      "#ff0000",
      "#00ff00",
      "#0000ff",
      "#ffff00",
      "#00ffff",
      "#ff00ff",
    ]
  }
  onChange={[Function]}
  value="#ff0000"
>
  <div
    className="color-picker"
  >
    <div
      className="clearfix toggle-wrap"
    >
      <ColorButton
        color="#ff0000"
        disabled={false}
        isSelected={false}
        onClick={[Function]}
      >
        <div
          className="color-button"
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            <button
              className="btn btn-default"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <div
                className="color-tile"
                style={
                  Object {
                    "backgroundColor": "#ff0000",
                  }
                }
              />
            </button>
          </Button>
        </div>
      </ColorButton>
    </div>
    <Overlay
      animation={[Function]}
      onHide={[Function]}
      placement="bottom"
      rootClose={true}
      show={false}
      target={[Function]}
    >
      <Overlay
        onHide={[Function]}
        placement="bottom"
        rootClose={true}
        show={false}
        target={[Function]}
        transition={[Function]}
      />
    </Overlay>
  </div>
</ColorPicker>
`;
